aten/src/ATen/native/quantized/cpu/qnnpack/test/gemm-microkernel-tester.h

```diff
@@ -310,6 +310,126 @@ class GemmMicrokernelTester {
+    std::vector<float> c((m() - 1) * cStride() + n());
+    std::vector<int32_t> acc(m() * n());
+
+    const uint8_t* aPtr = a.data() + 8;
+
+    for (size_t iteration = 0; iteration < iterations(); iteration++) {
+      std::generate(a.begin(), a.end(), std::ref(u8rng));
+      std::generate(b.begin(), b.end(), std::ref(u8rng));
+      std::generate(bias.begin(), bias.end(), std::ref(s32rng));
+      std::fill(c.begin(), c.end(), 0.0f);
+
+      std::fill(packedW.begin(), packedW.end(), bZeroPoint());
+
+      pytorch_pack_q8gemm_w(
+          n(),
+          k(),
+          nr(),
+          np(),
+          kr(),
+#if !PYTORCH_QNNPACK_RUNTIME_QUANTIZATION
+          aZeroPoint(),
+          bZeroPoint(),
+#endif
+          b.data(),
+          bias.data(),
+          packedW.data());
+
+      ASSERT_NE(
+          *std::max_element(a.cbegin(), a.cend()),
+          *std::min_element(a.cbegin(), a.cend()));
+      ASSERT_NE(
+          *std::max_element(b.cbegin(), b.cend()),
+          *std::min_element(b.cbegin(), b.cend()));
+
+      /* Compute 32-bit results and output quantization arguments */
+      std::fill(acc.begin(), acc.end(), 0);
+      for (size_t mIndex = 0; mIndex < m(); mIndex++) {
+        for (size_t nIndex = 0; nIndex < n(); nIndex++) {
+          for (size_t kIndex = 0; kIndex < k(); kIndex++) {
+            ASSERT_LE(n(), packedN());
+            ASSERT_LT(mIndex * n() + nIndex, acc.size());
+            ASSERT_LT(mIndex * k() + kIndex, a.size());
+            acc[mIndex * n() + nIndex] +=
+                (int32_t(aPtr[mIndex * aStride() + kIndex]) -
+                 int32_t(aZeroPoint())) *
+                (int32_t(b[nIndex * k() + kIndex]) - int32_t(bZeroPoint()));
+          }
+          acc[mIndex * n() + nIndex] += bias[nIndex];
+        }
+      }
+
+      const int32_t accMin = *std::min_element(acc.cbegin(), acc.cend());
+      const int32_t accMax = *std::max_element(acc.cbegin(), acc.cend());
+      if (m() * n() >= 3) {
+        ASSERT_NE(accMax, accMin)
+            << "Mr x Nr x Kr = " << mr() << " x " << nr() << " x " << kr()
+            << ", M x N x K = " << m() << " x " << n() << " x " << k();
+      }
+
+      const double cScale = uint32_t(accMax - accMin) >= 256
+          ? double(uint32_t(accMax - accMin)) / 255.0
+          : 1.00001;
+      const uint8_t cZeroPoint = uint8_t(std::max(
+          std::min(
+              lrint(127.5 - 0.5 * double(accMin + accMax) / cScale),
+              long(std::numeric_limits<uint8_t>::max())),
+          long(std::numeric_limits<uint8_t>::min())));
+
+      const float requantizationScale = 1.0f / float(cScale);
+      const union pytorch_qnnp_conv_quantization_params quantizationParams =
+          pytorch_qnnp_compute_conv_quantization_params(
+              aZeroPoint(),
+              bZeroPoint(),
+              requantizationScale,
+              cZeroPoint,
+              qmin(),
+              qmax());
```

Comment:
@supriyar I still need to calculate pytorch_qnnp_conv_quantization_params even though only the input and kernel zero points are getting used.  Please let me know what API works best for you.  Maybe it is just enough to pass in the input and kernel zero points and the multiplier as three arguments.  Let me know what you prefer.

Reply:
Yes, I think having the pytorch_qnnp_conv_quantization_params is an overkill. We can just pass in the zero_points and the multiplier as arguments to the kernel directly.